classes/observation_manager.php

```diff
@@ -0,0 +1,222 @@
+     * @param string $tablename database table name
+     * @return object existing point data
+     */
+    public static function get_existing_point_data(int $observationid, int $pointid,
+        string $tablename = 'observation_points'): object {
+
+        global $DB;
+        return $DB->get_record($tablename, ['id' => $pointid, 'obs_id' => $observationid], '*', MUST_EXIST);
+    }
+
+    /**
+     * Get all observation points in an observation instance
+     * @param int $observationid ID of the observation instance
+     * @param string $sortby column to sort by
+     * @param string $tablename database tablename
+     * @return array array of database objects obtained from database
+     */
+    public static function get_observation_points(int $observationid, string $sortby='list_order',
+        string $tablename='observation_points'): array {
+
+        global $DB;
+        return $DB->get_records($tablename, ['obs_id' => $observationid], $sortby);
+    }
+
+    /**
+     * Deletes observation point
+     * @param int $observationid ID of the observation instance
+     * @param int $obpointid ID of the observation point to delete
+     * @param string $tablename database table name
+     */
+    public static function delete_observation_point(int $observationid, int $obpointid, string $tablename='observation_points') {
+        global $DB;
+        // To ensure ordering stays intact, should move all those points with a higher order than this one down by 1.
+        $currentpoint = self::get_existing_point_data($observationid, $obpointid, $tablename);
+
+        // Get those with a higher list ordering than this one.
+        $pointsabove = $DB->get_records_select(
+            $tablename,
+            "obs_id = :obsid AND list_order > :listorder",
+            [
+                'obsid' => $observationid,
+                'listorder' => $currentpoint->list_order
+            ]
+        );
+
+        $transaction = $DB->start_delegated_transaction();
+
+        $DB->delete_records($tablename, ['id' => $obpointid, 'obs_id' => $observationid]);
+
+        // Shuffle those above down.
+        foreach ($pointsabove as $pointabove) {
+            $DB->update_record(
+                $tablename,
+                [
+                    'id' => $pointabove->id,
+                    'list_order' => $pointabove->list_order - 1
+                ]
+            );
+        }
+
+        $transaction->allow_commit();
+    }
+
+    /**
+     * Reorders an observation point in relation to the other observation points in this observation instance
+     * @param int $observationid ID of the observation instance
+     * @param int $obpointid ID of the observation point to reorder
+     * @param int $direction direction to reorder the point in, must be -1 or 1.
+     * @param string $tablename database table name
+     */
+    public static function reorder_observation_point(int $observationid, int $obpointid, int $direction,
```

Comment:
is an int the best data type to handle the direction? A bool seems like the more obvious choice, true => positive, false => negative. That way you dont need to a check around the value, the type hinting will exception if its not a bool, and both are valid values

Reply:
The idea with the integers instead of bools is that if we were to implement some sort of drag and drop reordering, it would be simple to refactor this function to reorder a point by more than 1 in either direction, however as it is implemented now that isn't supported so it is locked to -1 and 1. I may refactor this to allow and test for reordering by more than 1 space, as it shouldn't be too hard.